promql/prettier/prettier.go

```diff
@@ -0,0 +1,533 @@
+		}
+	case parser.RIGHT_PAREN:
+		if ((nodeSplittable && !p.pd.containsGrouping) || p.pd.multiArgumentCall) && !p.pd.expectAggregationLabels {
+			result += p.pd.newLine() + p.pd.pad(headInfo.baseIndent(item))
+		}
+		result += item.Val
+		if p.pd.containsGrouping {
+			p.pd.containsGrouping = false
+			result += p.pd.newLine()
+		}
+		if p.pd.expectAggregationLabels {
+			p.pd.expectAggregationLabels = false
+			result += " "
+		}
+	case parser.LEFT_BRACE:
+		result += item.Val
+		if nodeSplittable {
+			result += p.pd.newLine()
+			p.pd.insideMultilineBraces = true
+		}
+	case parser.RIGHT_BRACE:
+		if p.pd.insideMultilineBraces {
+			if items[index-1].Typ != parser.COMMA {
+				// Edge-case: if the labels are multi-line split, but do not have
+				// a pre-applied comma.
+				result += "," + p.pd.newLine() + p.pd.pad(headInfo.baseIndent(items[index]))
+			}
+			p.pd.insideMultilineBraces = false
+			p.pd.isNewLineApplied = false
+		}
+		result += item.Val
+	case parser.IDENTIFIER:
+		if p.pd.insideMultilineBraces {
+			result += p.pd.pad(1) + item.Val
+		} else {
+			result += item.Val
+		}
+	case parser.STRING, parser.NUMBER:
+		result += item.Val
+	case parser.SUM, parser.BOTTOMK, parser.COUNT_VALUES, parser.COUNT, parser.GROUP, parser.MAX, parser.MIN,
+		parser.QUANTILE, parser.STDVAR, parser.STDDEV, parser.TOPK:
+		// Aggregations.
+		result += item.Val + " "
+	case parser.EQL, parser.EQL_REGEX, parser.NEQ, parser.NEQ_REGEX, parser.DURATION, parser.COLON,
+		parser.LEFT_BRACKET, parser.RIGHT_BRACKET:
+		// Comparision operators.
+		result += item.Val
+	case parser.ADD, parser.SUB, parser.DIV, parser.GTE, parser.GTR, parser.LOR, parser.LAND,
+		parser.LSS, parser.LTE, parser.LUNLESS, parser.MOD, parser.POW:
+		// Vector matching operators.
+		if p.pd.containsGrouping {
+			result += p.pd.newLine() + p.pd.pad(baseIndent) + item.Val + " "
+			p.pd.isNewLineApplied = false
+		} else if nodeSplittable && !p.pd.immediateScalar {
+			result += p.pd.newLine() + p.pd.pad(baseIndent) + item.Val + p.pd.newLine()
+		} else {
+			result += " " + item.Val + " "
+		}
+		if p.pd.immediateScalar {
+			p.pd.immediateScalar = false
+		}
+	case parser.WITHOUT, parser.BY, parser.IGNORING, parser.BOOL, parser.GROUP_LEFT, parser.GROUP_RIGHT,
+		parser.OFFSET, parser.ON:
+		// Keywords.
+		result += item.Val
+		if item.Typ == parser.BOOL {
+			result += p.pd.newLine()
+		} else if p.pd.isAggregation {
+			p.pd.expectAggregationLabels = true
+		}
+	case parser.COMMA:
+		result += item.Val
+		if nodeSplittable || p.pd.multiArgumentCall {
+			result += p.pd.newLine()
+		} else {
+			result += " "
+		}
+	case parser.COMMENT:
+		if result[len(result)-1] != ' ' {
+			result += " "
+		}
+		if p.pd.isNewLineApplied {
+			result += p.pd.pad(currNode.previousIndent()) + item.Val
+		} else {
+			result += item.Val
+		}
+		p.pd.isPreviousItemComment = true
+	}
+	if index+1 == len(items) {
+		return result, nil
+	}
+	ptmp, err := p.prettify(items, index+1, result)
```

Comment:
Here, a loop might be better than tail recursion.

Reply:
Yes it will. I am planning to replace it with an iterator. I hope, that will be even better.

Reply:
I have implemented the `itemsIterator`. Please have a look over it.